### PR TITLE
Implement registration framework

### PR DESCRIPTION
**Pull request by tbirdso**
Implements the `itk-dreg` out-of-memory registration framework. Includes installable modules, testing, and examples.

Refer to updated README and CONTRIBUTING documentation along with notebook examples to get started.

Deferred: Deferred to https://github.com/InsightSoftwareConsortium/itk-dreg/issues/7: Run distributed with `LocalCluster`. Depends on https://github.com/InsightSoftwareConsortium/ITK/issues/4267.

cc @thewtex

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,99 @@
...
+```py
+my_fixed_image = ...
+my_moving_image = ...
+
+my_initial_transform = ...
+
+# registration method returns an update to the initial transform
+
+my_transform_delayed = itk_dreg.register_images(
+    target_da=target_image_dask_voxel_array,
```
Comment (thewtex) [inline README.md:37]: Can this be obtained from the reader?

File: README.md [hunk 2]
```diff
@@ -2,6 +2,99 @@
...
+    initial_transform=my_initial_transform,
+    source_reader_ctor=my_construct_streaming_reader_method,
+    target_reader_ctor=my_construct_streaming_reader_method,
```
Comment (thewtex) [inline README.md:40]: source, target -> moving, fixed

File: docs/develop.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+dask.config.set(scheduler='single-threaded')
+```
+
+The `pdb` Python module can be useful to set breakpoints and examine state in a local
+Python environment.
+
+```
+... # your code
+import pdb
+pdb.set_trace() # halts execution for interactive debugging
```
Comment (thewtex) [inline docs/develop.md:27]: These two can be reduced in modern Python with

```
breakpoint()
```

File: docs/develop.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+`dask.distributed` supports the Python `logging` library as the preferred method for
+collecting log information from workers. Logging behavior can be configured
+in the Dask `Client` instance. See Dask
+[logging documentation](https://distributed.dask.org/en/latest/logging.html)
+for more information on event-based logging.
+
+- We strongly encourage developers to write unit and integration tests during development
+to help identify failure points and to maintain quality code practices. `itk_dreg` uses Pytest
+for its testing infrastructure.
```
Comment (thewtex) [inline docs/develop.md:44]: :+1: 

Pytest -> pytest

File: docs/develop.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+- Remote debugging: Dask [debugging documentation](https://docs.dask.org/en/latest/how-to/debug.html)
+discusses approaches for parallel and remote debugging when running
+on a cluster, including printing with `dask.distributed.print` and exception
+handling.
+
+`dask.distributed` supports the Python `logging` library as the preferred method for
+collecting log information from workers. Logging behavior can be configured
+in the Dask `Client` instance. See Dask
+[logging documentation](https://distributed.dask.org/en/latest/logging.html)
+for more information on event-based logging.
```
Comment (thewtex) [inline docs/develop.md:40]: :+1:

**Review by thewtex — COMMENTED** [review]
:rocket:

File: README.md (hunk 1 above)
Comment (tbirdso) [reply README.md:37]: Yes, though I want to ensure that the user can understand how the chunk registration schedule is composed before registration actually runs.

In your view, would it be reasonable to update `register_images` to accept a `chunk_size` and return the corresponding  `dask.array.Array` representation alongside the delayed task graph?

**Comment by thewtex** [comment]
stitch_reduce -> displacement_field_reduce

**Comment by thewtex** [comment]
dask array argument -> chunk size specification that is used for internal dask array construction

**Comment by thewtex** [comment]
Add a comment noting that the current readers / writers will work with any itk-supported image type, but streaming will require a streaming supported reader, e.g. uncompressed MetaImage or Zarr.

**Comment by tbirdso** [comment]
@thewtex Following up from discussion yesterday, documentation on the Dask "threads" scheduler is here: https://docs.dask.org/en/stable/scheduling.html#local-threads

Parallelism occurs within a single process. I have not run benchmarks to determine whether we are seeing speedup with ITKElastix between "synchronous" and "threads" scheduling.

**Comment by tbirdso** [comment]
@thewtex Ping: this PR is ready for review and then merge. Additions are significant, I leave it up to you to decide how closely you'd like to do code review following our previous walkthroughs together offline.

**Review by thewtex — APPROVED** [review]
@tbirdso Phenomenal! :1st_place_medal: :champagne: :sparkler: 

I love the tests and CI!

**Pull request merged**